core/src/main/java/org/infinispan/remoting/inboundhandler/BaseBlockingRunnable.java

```diff
@@ -28,10 +28,18 @@ protected BaseBlockingRunnable(BasePerCacheInboundInvocationHandler handler, Cac
 
    @Override
    public void run() {
+      InvocationStatus invocationStatus = null;
       try {
-         response = beforeInvoke();
-         if (response == null) {
-            response = handler.invokePerform(command);
+         invocationStatus = beforeInvoke();
+         switch (invocationStatus) {
+            case CONTINUE:
+               response = handler.invokePerform(command);
+               break;
+            case CACHE_NOT_FOUND:
+               response = CacheNotFoundResponse.INSTANCE;
+               break;
+            case DEFERRED:
+               return;
```

Comment:
@rvansa To answer your first question, yes, the way I did this in https://github.com/infinispan/infinispan/pull/4552 was to add a `sync` field to `BaseBlockingRunnable`: 
https://github.com/infinispan/infinispan/pull/4552/files#diff-f117a0bc3504f81ca8b7036425f73a6eR28

Now back to the more important issue of `DIST_ASYNC` :)
IMO with https://github.com/infinispan/infinispan/pull/4552, the main perf difference between `DIST_ASYNC` and `putAsync()` will be the fact that the primary releases the lock immediately after sending the command to the backups, so you can have a lot more concurrent writes to the same key. You're right though, the difference will be much smaller than it is now, so perhaps we *should* implement `DIST_ASYNC` as "replace all calls to put() with putAsync()" in 9.0. Perhaps I should start thread on the dev list :)